---
title: "Class 10: Structural BioInformatics II"
author: "Sharlene Yang (PID: A16351303)"
format: pdf
---

# Comparative Analysis of ADK

ADK (adenelate Kinase) is an important drug target and we would love to know how it workd - i.e. molecular mechanism

There has been lots of work done on this protein due to it's importance inclusing lots of crystal structures.

We will begin getting an example ADK sequence from the database.
We will then use this to find all ADK structures in the PDB.

```{r}
library(bio3d)
aa <- get.seq("1ake_A")
aa
```

> Q10. Which of the packages above is found only on BioConductor and not CRAN?

"msa"

> Q11. Which of the above packages is not found on BioConductor or CRAN?:

"bio3d-view"

>Q12. True or False? Functions from the devtools package can be used to install packages from GitHub and BitBucket? 

TRUE

>Q13. How many amino acids are in this sequence, i.e. how long is this sequence?

214 amino acids

We can now run BLAST with this sequence
```{r}
# Blast or hmmer search 
#b <- blast.pdb(aa)
```

```{r}
# Plot a summary of search results
#hits <- plot(b)
```

NOTE: -log, flips the axis so the important data is on the top. In this case, since log is negative, we want our E-value to be bigger. 

Let's see what is in out `hits` objects
```{r}
# List out some 'top hits'
#head(hits$pdb.id)

hits <- NULL
hits$pdb.id <- c('1AKE_A','6S36_A','6RZE_A','3HPR_A','1E4V_A','5EJE_A','1E4Y_A','3X2S_A','6HAP_A','6HAM_A','4K46_A','3GMT_A','4PZL_A')
```

Now we can download all these PDB structure files:

```{r}
# Download related PDB files
files <- get.pdb(hits$pdb.id, path="pdbs", split=TRUE, gzip=TRUE)

```


Now I want to align and superpose these structures which are all over the place.

```{r}
# Align related PDBs
pdbs <- pdbaln(files, fit = TRUE, exefile="msa")
```

Let's have a look at out `pdbs` object:
```{r}
pdbs
```

Now we have our aligned and superposed structures we can perform all sorts of analysis on them. Let's do PCA...

```{r}
# Perform PCA
pc.xray <- pca(pdbs)
plot(pc.xray)
```

Results of PCA on Adenylate kinase X-ray structures. Each dot represents one PDB structure.

We can cluster the structures by RMSD (or any other method). Function `rmsd()` will calculate all pairwise RMSD values of the structural ensemble. This facilitates clustering analysis based on the pairwise structural deviation:

```{r}
# Calculate RMSD
rd <- rmsd(pdbs)

# Structure-based clustering
hc.rd <- hclust(dist(rd))
grps <- cutree(hc.rd, k=3)

plot(pc.xray, 1:2, col=grps)
```



## Optional further visualization

We can make a wee movie - also called a trajectory of the major differences (i.e. structural displacements) of ADK.

```{r}
# Visualize first principal component
pc1 <- mktrj(pc.xray, pc=1, file="pc_1.pdb")
pc1
```


